src/attribution/onINP.ts

```diff
@@ -25,38 +26,109 @@ import {
+const MAX_LOAFS_TO_CONSIDER = 10;
+
+// A list of longest LoAFs with interactions on the page sorted so the
+// longest one is first. The list is at most MAX_LOAFS_TO_CONSIDER long.
+const longestLoAFsList: PerformanceLongAnimationFrameTiming[] = [];
+
+// A PerformanceObserver, observing new `long-animation-frame` entries.
+// If this variable is defined it means the browser supports LoAF.
+let loafObserver: PerformanceObserver | undefined;
+
+const handleEntries = (entries: PerformanceLongAnimationFrameTiming[]) => {
+  entries.forEach((entry) => {
+    if (entry.firstUIEventTimestamp > 0) {
+      const minLongestLoAF = longestLoAFsList[longestLoAFsList.length - 1];
+
+      if (!longestLoAFsList.length) {
+        longestLoAFsList.push(entry);
+        return;
+      }
+
+      // If the entry is possibly one of the 10 longest, insert it into the
+      // list of pending LoAFs at the correct spot (sorted), ensuring the list
+      // is not longer than `MAX_LOAFS_TO_CONSIDER`.
+      if (
+        longestLoAFsList.length < MAX_LOAFS_TO_CONSIDER ||
+        entry.duration > minLongestLoAF.duration
+      ) {
+        for (let i = 0; i < longestLoAFsList.length; i++) {
+          if (entry.duration > longestLoAFsList[i].duration) {
+            longestLoAFsList.splice(i, 0, entry);
+            break;
+          }
+        }
+        if (longestLoAFsList.length > MAX_LOAFS_TO_CONSIDER) {
+          longestLoAFsList.splice(MAX_LOAFS_TO_CONSIDER);
+        }
+      }
+    }
+  });
+};
+
 const attributeINP = (metric: INPMetric): void => {
-  if (metric.entries.length) {
-    const longestEntry = metric.entries.sort((a, b) => {
-      // Sort by: 1) duration (DESC), then 2) processing time (DESC)
-      return (
-        b.duration - a.duration ||
-        b.processingEnd -
-          b.processingStart -
-          (a.processingEnd - a.processingStart)
-      );
-    })[0];
-
-    // Currently Chrome can return a null target for certain event types
-    // (especially pointer events). As the event target should be the same
-    // for all events in the same interaction, we pick the first non-null one.
-    // TODO: remove when 1367329 is resolved
-    // https://bugs.chromium.org/p/chromium/issues/detail?id=1367329
-    const firstEntryWithTarget = metric.entries.find((entry) => entry.target);
-
-    (metric as INPMetricWithAttribution).attribution = {
-      eventTarget: getSelector(
-        firstEntryWithTarget && firstEntryWithTarget.target,
-      ),
-      eventType: longestEntry.name,
-      eventTime: longestEntry.startTime,
-      eventEntry: longestEntry,
-      loadState: getLoadState(longestEntry.startTime),
-    };
-    return;
+  const sortedEntries = metric.entries.sort((a, b) => {
+    return a.processingStart - b.processingStart;
+  });
+
+  const firstEntry = sortedEntries[0];
+  const lastEntry = sortedEntries[sortedEntries.length - 1];
+  const renderTime = firstEntry.startTime + firstEntry.duration;
```

Comment:
...I also like to take Math.min(renderTime, lastEntry.processingEnd), just in case.

Specifically, for cases where an interaction does NOT have a paint but still had a long task-- then you have a 50/50 chance that duration was rounded down and now  renderTime appears less than processingTime.